api/v1alpha1/virtualmachine_conversion.go

```diff
@@ -892,13 +892,150 @@ func restore_v1alpha2_VirtualMachineReadinessProbeSpec(
+	var preReqCondIdx int
+
+	for i := range dst.Status.Conditions {
+		c := &dst.Status.Conditions[i]
+
+		switch c.Type {
+		case string(VirtualMachinePrereqReadyCondition):
+			preReqCond = c
+			preReqCondIdx = i
+		case v1alpha2.VirtualMachineConditionClassReady:
+			vmClassCond = c
+		case v1alpha2.VirtualMachineConditionImageReady:
+			vmImageCond = c
+		case v1alpha2.VirtualMachineConditionVMSetResourcePolicyReady:
+			vmSetResourcePolicy = c
+		case v1alpha2.VirtualMachineConditionBootstrapReady:
+			vmBootstrap = c
+		}
+	}
+
+	// If we didn't find the v1a1 PrereqReady condition then there is nothing to do.
+	if preReqCond == nil {
+		return dst.Status.Conditions
+	}
+
+	// Remove the v1a1 PrereqReady condition if not nil
+	dst.Status.Conditions = append(dst.Status.Conditions[:preReqCondIdx], dst.Status.Conditions[preReqCondIdx+1:]...)
+
+	// If any of the v1a2 conditions were already evaluated, v1a1 PrereqReady condition has been removed, so move on.
+	if vmClassCond != nil || vmImageCond != nil || vmSetResourcePolicy != nil || vmBootstrap != nil {
+		return dst.Status.Conditions
+	}
+
+	var conditions []metav1.Condition
+	// If we don't have any of the new v1a2 conditions, use the v1a1 PrereqReady condition to fill
+	// that in. This means that this was originally a v1a1 VM.
+	if preReqCond.Status == metav1.ConditionTrue {
+
+		// The class and image are always required fields.
+		conditions = append(conditions, metav1.Condition{
+			Type:               v1alpha2.VirtualMachineConditionClassReady,
+			Status:             metav1.ConditionTrue,
+			LastTransitionTime: preReqCond.LastTransitionTime,
+			Reason:             string(metav1.ConditionTrue),
+		})
+		conditions = append(conditions, metav1.Condition{
+			Type:               v1alpha2.VirtualMachineConditionImageReady,
+			Status:             metav1.ConditionTrue,
+			LastTransitionTime: preReqCond.LastTransitionTime,
+			Reason:             string(metav1.ConditionTrue),
+		})
+
+		if dst.Spec.Reserved != nil && dst.Spec.Reserved.ResourcePolicyName != "" {
+			conditions = append(conditions, metav1.Condition{
+				Type:               v1alpha2.VirtualMachineConditionVMSetResourcePolicyReady,
+				Status:             metav1.ConditionTrue,
+				LastTransitionTime: preReqCond.LastTransitionTime,
+				Reason:             string(metav1.ConditionTrue),
+			})
+		}
+		if dst.Spec.Bootstrap != nil {
+			conditions = append(conditions, metav1.Condition{
+				Type:               v1alpha2.VirtualMachineConditionBootstrapReady,
+				Status:             metav1.ConditionTrue,
+				LastTransitionTime: preReqCond.LastTransitionTime,
+				Reason:             string(metav1.ConditionTrue),
+			})
+		}
+
+	} else if preReqCond.Status == metav1.ConditionFalse {
+		// Infer what v1a2 conditions need to be true/false from preReq reason.
+		// Order of evaluation of objects for v1a1 PreReq Condition:
+		// 1. VM Class then
+		// 2. VM Image
+
+		if preReqCond.Reason == VirtualMachineClassNotFoundReason {
```

Comment:
Not to likely to happend but I think could depending on what version upgrade from, we also include the `VirtualMachineClassBindingNotFoundReason` here? And use the PrereqReady reason as the new Reason too.

Reply:
Just to confirm -- Do we wanna set the PreReqReady reason as the new reason for the v1a2 Conditions too?. Notice I've used False/True which are used as v1a2 reasons for VMClass/Image not Ready otherwise.